src/interp.rs

```diff
@@ -19,153 +19,483 @@
+        self[1] += other[1] * coeff;
+        self[2] += other[2] * coeff;
+        self[3] += other[3] * coeff;
+        self[4] += other[4] * coeff;
+        self[5] += other[5] * coeff;
+    }
+
+    #[inline(always)]
+    fn from_planar(slice: &[f32], stride: usize) -> Self {
+        [
+            slice[0],
+            slice[stride],
+            slice[2 * stride],
+            slice[3 * stride],
+            slice[4 * stride],
+            slice[5 * stride],
+        ]
+    }
+}
+
+mod specialized {
+    use super::Frame;
+    use std::f64::consts::PI;
+
+    const ALMOST_ZERO: f64 = 0.000001;
+    const TAPS: usize = 48;
+
+    const FACTOR4: usize = 4;
+    const FACTOR4_INPUT_LENGTH: usize = TAPS / FACTOR4;
+    const FACTOR2: usize = 2;
+    const FACTOR2_INPUT_LENGTH: usize = TAPS / FACTOR2;
+
+    #[derive(Debug)]
+    pub(super) struct Interp4F<F: Frame> {
+        filter: [[f32; FACTOR4]; FACTOR4_INPUT_LENGTH],
+        buffer: [F; FACTOR4_INPUT_LENGTH],
+        buffer_pos: usize,
+    }
+
+    impl<'a, 'b, F> Interp4F<F>
+    where
+        F: Frame + Default + AsRef<[f32]> + 'static,
+    {
+        pub(super) fn new() -> Self {
+            let mut filter: [[_; FACTOR4]; FACTOR4_INPUT_LENGTH] = Default::default();
+            for (j, coeff) in filter
+                .iter_mut()
+                .map(|x| x.iter_mut())
+                .flatten()
+                .enumerate()
+            {
+                let j = j as f64;
+                // Calculate Hanning window, with one tap ignored. (Last tap is zero anyways, and we want to hit
+                // an even multiple of 48)
+                let window = (TAPS - 1 + 1) as f64;
+                let w = 0.5 * (1.0 - f64::cos(2.0 * PI * j / window));
+
+                // Calculate sinc and apply hanning window
+                let m = j - window / 2.0;
+                *coeff = if m.abs() > ALMOST_ZERO {
+                    w * f64::sin(m * PI / FACTOR4 as f64) / (m * PI / FACTOR4 as f64)
+                } else {
+                    w
+                } as f32;
+            }
+
+            Self {
+                filter,
+                buffer: Default::default(),
+                buffer_pos: (FACTOR4_INPUT_LENGTH) - 1,
+            }
+        }
+
+        pub(super) fn reset(&mut self) {
+            self.buffer = Default::default();
+        }
+
+        pub(super) fn push(&mut self, frame: &F) -> [F; FACTOR4] {
+            // Write in Frames in reverse, to enable forward-scanning with filter
+            self.buffer_pos = (self.buffer_pos + self.buffer.len() - 1) % self.buffer.len();
+            self.buffer[self.buffer_pos] = *frame;
+
+            let mut output: [F; FACTOR4] = Default::default();
+
+            let mut filterp = 0;
+
+            for input_frame in &self.buffer[self.buffer_pos..] {
+                let filter_coeffs = &self.filter[filterp];
+                for (output_frame, coeff) in Iterator::zip(output.iter_mut(), filter_coeffs) {
```

Comment:
```suggestion
                for (output_frame, coeff) in output.iter_mut().zip(filter_coeffs) {
```

Why not that? But now that I read yours, using `zip()` not as an associated method seems more readable :)

Reply:
Only since I too found it more readable. :) (I'll keep it then)

Reply:
See also https://github.com/sdroege/ebur128/issues/35